### PR TITLE
feat(RUN-1032): Parallel dirty page copying for page allocator

### PR DESCRIPTION
**Pull request by alexandru-uta**
For messages that dirty large amounts of pages, speedups can be achieved by copying the dirty pages in parallel at the end of message execution. This MR parallelizes dirty page copying for messages that dirty more than 64 MB.

### CODE REVIEW

File: rs/config/src/embedders.rs [hunk 1]
```diff
@@ -39,6 +39,9 @@ pub(crate) const DEFAULT_COST_TO_COMPILE_WASM_INSTRUCTION: NumInstructions =
 /// The number of rayon threads used by wasmtime to compile wasm binaries
 const DEFAULT_WASMTIME_RAYON_COMPILATION_THREADS: usize = 10;
 
+/// The number of rayon threads use for the parallel page copying optimization.
+const DEFAULT_PAGE_ALLOCATOR_THREADS: usize = 8;
```
Comment (schneiderstefan) [inline rs/config/src/embedders.rs:43]: Out of curiosity, how are these numbers of 8 and 10 chosen?

File: rs/replicated_state/Cargo.toml [hunk 2]
```diff
@@ -36,6 +36,7 @@ maplit = "1.0.2"
 nix = { workspace = true }
 rand = { workspace = true }
 rand_chacha = { workspace = true }
+rayon = "1.5.1"
```
Comment (schneiderstefan) [inline rs/replicated_state/Cargo.toml:39]: Can we use `workspace = true` here? As far as I know this is the preferred way of specifying versions.

File: rs/replicated_state/src/page_map/page_allocator.rs [hunk 3]
```diff
@@ -84,6 +89,11 @@ impl PageAllocator {
     /// iterator. Knowing the page count beforehand allows the page allocator
     /// to optimize allocation.
     pub fn allocate(&self, pages: &[(PageIndex, &PageBytes)]) -> Vec<(PageIndex, Page)> {
+        // If the pages that need to be allocated and copied are more than 100 MB,
```
Comment (schneiderstefan) [inline rs/replicated_state/src/page_map/page_allocator.rs]: 100MB?

**Review by schneiderstefan — APPROVED** [review]

File: rs/replicated_state/src/page_map/page_allocator.rs (hunk 3 above)
Comment (schneiderstefan) [reply rs/replicated_state/src/page_map/page_allocator.rs]: Sorry was looking at an old version.

File: rs/config/src/embedders.rs (hunk 1 above)
Comment (alexandru-uta) [reply rs/config/src/embedders.rs:43]: They're both chosen from benchmarks. Both are "sweetspots" for the two operations. If we increase the numbers some more we see some diminishing returns.

File: rs/replicated_state/src/page_map/page_allocator.rs (hunk 3 above)
Comment (alexandru-uta) [reply rs/replicated_state/src/page_map/page_allocator.rs]: yes, thanks, I noticed that as well. Fixed now.

File: rs/replicated_state/Cargo.toml (hunk 2 above)
Comment (alexandru-uta) [reply rs/replicated_state/Cargo.toml]: Thanks, fixed now.

**Comment by berestovskyy** [comment]
Could you please point to a relevant benchmark or include in this MR?

**Comment by alexandru-uta** [comment]
> Could you please point to a relevant benchmark or include it in this MR?

Screenshot 2024-08-06 at 09 15 32

That's what I had when I did the performance investigation.

**Comment by berestovskyy** [comment]
Could you please include your benchmark in this MR, similar to the benchmarks here: https://github.com/dfinity/ic/tree/master/rs/embedders/benches

**Comment by alexandru-uta** [comment]
> Could you please include your benchmark in this MR, similar to the benchmarks here: https://github.com/dfinity/ic/tree/master/rs/embedders/benches

Let's do that in a subsequent PR.

**Review by berestovskyy — APPROVED** [review]

**Pull request merged**